app/scripts/metamask-controller.js

```diff
@@ -6324,6 +6325,168 @@ export default class MetamaskController extends EventEmitter {
     return this.sortEvmAccountsByLastSelected(ethAccounts);
   }
 
+  /**
+   * Handles Hyperliquid referral approval flow for eth_accounts requests.
+   * Shows approval confirmation screen if needed and manages referral URL redirection.
+   *
+   * @param {object} req - The JSON-RPC request object.
+   */
+  async handleHyperliquidReferral(req) {
+    const { origin, tabId } = req;
+    const HYPERLIQUID_ORIGIN = 'https://app.hyperliquid.xyz';
+
+    const isHyperliquidReferralEnabled =
+      this.remoteFeatureFlagController?.state?.remoteFeatureFlags
+        ?.extensionUxDefiReferral;
+
+    if (origin !== HYPERLIQUID_ORIGIN || !isHyperliquidReferralEnabled) {
+      return;
+    }
+
+    const permittedAccounts = this.getPermittedAccounts(origin);
+
+    // Only show consent if Hyperliquid has permitted accounts
+    if (permittedAccounts.length === 0) {
+      return;
+    }
+
+    const permittedAccount = permittedAccounts[0];
+
+    // Check if there's already a pending approval request to prevent duplicates
+    // TODO: could be a problem if other approvals are shown on this url?
+    if (this.approvalController.has({ origin })) {
+      return;
+    }
+
+    const {
+      approvedAccounts = [],
+      passedAccounts = [],
+      declinedAccounts = [],
+    } = this.preferencesController.state.referrals.hyperliquid;
+
+    const hasApproved = approvedAccounts.includes(permittedAccount);
+    const hasBeenRedirected = passedAccounts.includes(permittedAccount);
+    const hasDeclined = declinedAccounts.includes(permittedAccount);
+
+    // We should show approval screen if the account is not in any of the tracked states
+    const shouldShowApproval =
+      !hasApproved && !hasBeenRedirected && !hasDeclined;
+
+    // We should redirect to the referral url immediately
+    // if the account is approved but not passed or declined
+    const shouldRedirect = hasApproved && !hasBeenRedirected && !hasDeclined;
```

Comment:
hasApproved and hasDeclined should be opposites of each other, so only one should need to be checked. I think this highlights how this state would be better represented as `Record<hex, SomeNewReferralEnum>`